MathLib/mechanics_invariants.h

```diff
@@ -0,0 +1,166 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+#ifndef MATHLIB_MECHANICS_INVARIANTS_H
+#define MATHLIB_MECHANICS_INVARIANTS_H
+
+#include <Eigen/Dense>
+
+namespace MathLib
+{
+namespace Mechanics
+{
+/// Invariants used in mechanics, based on Kelvin representation of the vectors
+/// and matrices.
+/// The invariants are computed at process creation time.
+template <int KelvinVectorSize>
+struct Invariants final
+{
+    // Deviatoric projection matrix.
+    static Eigen::Matrix<double, KelvinVectorSize, KelvinVectorSize> const
+        P_dev;
+    // Spherical projection matrix.
+    static Eigen::Matrix<double, KelvinVectorSize, KelvinVectorSize> const
+        P_sph;
+    /// Kelvin mapping of 2nd order identity
+    static Eigen::Matrix<double, KelvinVectorSize, 1> const ivec;
+
+    // Determinant of a matrix in Kelvin vector representation.
+    static double determinant(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& vec);
+
+    // Inverse of a matrix in Kelvin vector representation.
+    static Eigen::Matrix<double, KelvinVectorSize, 1> inverse(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& vec);
+
+    static double effectiveStress(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return std::sqrt(3 * J2(deviatoric_stress));
+    }
+
+    static double J2(
+        Eigen::Matrix<double, KelvinVectorSize, 1> const& deviatoric_stress)
+    {
+        return 0.5 * deviatoric_stress.transpose() * deviatoric_stress;
```

Comment:
These get called quite often. So always checking tracelessness may not be efficient? In general you could pass any vector, make it deviatoric and then return the invariant. But in the interest of efficiency, I think the user "has to know" that deviatoric invariants need to be called with deviatoric arguments and would leave the implementation as is.